src/Common/ProxyConfigurationResolver.h

```diff
@@ -9,9 +9,41 @@ struct ProxyConfigurationResolver
 {
     using Protocol = ProxyConfiguration::Protocol;
 
+    enum class ConnectProtocolPolicy
+    {
+        DEFAULT,
+        FORCE_OFF,
+        FORCE_ON
+    };
+
+    explicit ProxyConfigurationResolver(Protocol request_protocol_, ConnectProtocolPolicy connect_protocol_policy_ = ConnectProtocolPolicy::DEFAULT)
+        : request_protocol(request_protocol_), connect_protocol_policy(connect_protocol_policy_)
+    {
+    }
+
     virtual ~ProxyConfigurationResolver() = default;
     virtual ProxyConfiguration resolve() = 0;
     virtual void errorReport(const ProxyConfiguration & config) = 0;
+
+protected:
+    Protocol request_protocol;
```

Comment:
I feel bad for introducing state in this class (which used to be a simple interface), but the request protocol as well as connect protocol policy is required to determine if connect should be on or off.

Reply:
This could actually be avoided for both environment resolver and remote resolver because the proxy protocol is known beforehand. When `ProxyConfigurationResolverProvider` is picking the configuration, it knows the proxy protocol and the request protocol, so it could simply inject a boolean.

The problem lies on `ProxyListConfigurationResolver` as it contains a list of URIs and that list might contain mixed proxy protocols. One thing I could do is have something like `std::vector<std::pair<URI, bool>>` instead of `std::vector<URI>` only for the list. That would simplify the other cases, but make this one a bit more complex.

Not sure it's worth it.

I am writing this comment (and code) with the feeling that there is a simpler way to address this